### PR TITLE
Adding deny all rule to Azure Bastion nsg

### PR DESCRIPTION
**Pull request by sid2305**
# Overview/Summary

[AB#26315](https://dev.azure.com/CSUSolEng/2a9d5a5a-8998-4ad0-81c8-ef3045e4da97/_workitems/edit/26315)

Added deny all rule to both inbound and outbound rules with the lowest priority i.e. 4096 for the Azure Bastion nsg and added parameters for destination ports of security rules. 

## This PR fixes/adds/changes/removes

1. adds deny all rule for inbound and outbound rules.
2. adds parameters with default values to hub networking parameter and bicep file for destination ports of inbound/outbound rules

## Testing Evidence

image

Replace this with any testing evidence to show that your Pull Request works/fixes as described and planned (include screenshots, if appropriate).

## As part of this Pull Request I have

- [ ] Read the [Contribution Guide](https://github.com/Azure/ALZ-Bicep/wiki/Contributing) and ensured this PR is compliant with the guide
- [ ] Ensured the resource API versions in `.bicep` file/s I am adding/editing are using the latest API version possible
- [ ] Checked for duplicate [Pull Requests](https://github.com/Azure/ALZ-Bicep/pulls)
- [ ] Associated it with relevant [GitHub Issues](https://github.com/Azure/ALZ-Bicep/issues)
- [ ] *(ALZ Bicep Core Team Only)* Associated it with relevant [ADO Items](https://aka.ms/alz/bicep/backlog)
- [ ] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/ALZ-Bicep/tree/main)
- [ ] Performed testing and provided evidence.
- [ ] Updated one or more of the following tests *(if required)*
  - [Unit](https://github.com/Azure/ALZ-Bicep/blob/main/.github/workflows/bicep-build-to-validate.yml)
  - [Linting](https://github.com/Azure/ALZ-Bicep/tree/main/.github/workflows)
  - [E2E (End-To-End)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/bicep-build-to-validate.yml)
  - [ValidateAzCloud (Base validation in Azure Cloud)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/base-unit-validate.yml)
  - [ValidateMcCloud (Base validation in Azure China Cloud)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/mc-base-unit-validate.yml)
- [ ] Updated relevant and associated documentation (e.g. Contribution Guide, Module READMEs, Wiki Docs etc.)
- [ ] If relevant, created or updated Code Tours [here](https://github.com/Azure/ALZ-Bicep/blob/main/.vscode/tours)

### CODE REVIEW

**Review by jtracey93 — CHANGES_REQUESTED** [review]
Hey @sid2305,

Looking good. Can you make sure your fork allows pushes from maintainers as currently the docs arent updating on your PR as seen here: https://github.com/Azure/ALZ-Bicep/actions/runs/4235275012/jobs/7358691021#step:11:160

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

Thanks

Jack

**Comment by sid2305** [comment]
Hi @jtracey93,
Turned on the option to allow pushes to this fork.

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep [hunk 1]
```diff
@@ -235,6 +235,30 @@ param parTags object = {}
...
+
+@sys.description('Define outbound azure cloud destination port range for Azure cloud. Default:443')
+param parAzureCloudPort string = '443'
+
+@sys.description('Define outbound destination port ranges for Bastion Communication. Default:[8080, 5701]')
+param parBastionCommunicationPorts array = ['8080','5701']
+
+@sys.description('Define outbound get session destination port range for Get Session Information. Default:80')
+param parGetSessionPort string = '80'
+
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep:261]: Can we make these names or descriptions a little more descriptive around what they are used fro. This module does an awful lot more than just deploy an NSG, so having something like `parHttpsPort` is to generic for this module.

For example `parHttpsPort` could/should be something like `parBastionInboundHttpsPort`

Please review and amend all of theses new parameters with this review comment in mind,

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep [hunk 2]
```diff
@@ -235,6 +235,30 @@ param parTags object = {}
 @sys.description('Set Parameter to true to Opt-out of deployment telemetry.')
 param parTelemetryOptOut bool = false
 
+@sys.description('Define inbound destination port or a range for HTTPS. Default:443')
+param parHttpsPort string = '443'
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep:239]: does this really need to be a parameter. Bastion is only accessible via 443 so therefore no customization option required.

Propose we remove and hardcode as before

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep [hunk 3]
```diff
@@ -235,6 +235,30 @@ param parTags object = {}
...
+
+@sys.description('Define inbound destination port or a range for gateway manager. Default:443')
+param parGatewayManagerPort string  ='443'
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep:242]: does this really need to be a parameter. Bastion gateway manager is only accessible via 443 so therefore no customization option required.

Propose we remove and hardcode as before

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep [hunk 4]
```diff
@@ -235,6 +235,30 @@ param parTags object = {}
...
+
+@sys.description('Define inbound destination port or a range for load balancer. Default:443')
+param parLoadBalancerPort string = '443'
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep:245]: Which load balancer? 

Propose we remove anyway and hardcode as before as this isnt something customers should be able to change

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep [hunk 5]
```diff
@@ -235,6 +235,30 @@ param parTags object = {}
...
+
+@sys.description('Define inbound destination ports or ranges for Bastion Host Communication. Default:[8080, 5701]')
+param parBastionHostCommunicationPorts array = ['8080','5701']
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep:248]: These ports cannot be changed on the bastion service, so they dont need to be exposed via a parameter. 

Propose we remove and hardcode

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep [hunk 6]
```diff
@@ -235,6 +235,30 @@ param parTags object = {}
...
+
+@sys.description('Define outbound destination ports or ranges for SshRDPOutbound. Default:[22,3389]')
+param parSshRDPPort array = ['22','3389']
+
+@sys.description('Define outbound azure cloud destination port range for Azure cloud. Default:443')
+param parAzureCloudPort string = '443'
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep:254]: This is a requirement for the bastion service to work, so lets not expose via parameter so customer cant break this by mistake.

Propose remove and hardcode as before

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep [hunk 7]
```diff
@@ -235,6 +235,30 @@ param parTags object = {}
...
+
+@sys.description('Define outbound destination port ranges for Bastion Communication. Default:[8080, 5701]')
+param parBastionCommunicationPorts array = ['8080','5701']
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep:257]: Duplicate + These ports cannot be changed on the bastion service, so they dont need to be exposed via a parameter.

Propose we remove and hardcode

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep [hunk 8]
```diff
@@ -235,6 +235,30 @@ param parTags object = {}
...
+param parGatewayManagerPort string  ='443'
+
+@sys.description('Define inbound destination port or a range for load balancer. Default:443')
+param parLoadBalancerPort string = '443'
+
+@sys.description('Define inbound destination ports or ranges for Bastion Host Communication. Default:[8080, 5701]')
+param parBastionHostCommunicationPorts array = ['8080','5701']
+
+@sys.description('Define outbound destination ports or ranges for SshRDPOutbound. Default:[22,3389]')
+param parSshRDPPort array = ['22','3389']
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep:251]: I think this is the only valid parameter to keep from my review

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep (hunk 1 above)
Comment (sid2305) [reply infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep]: Thanks for the review @jtracey93. It totally makes sense to only parameterize the SSH/RDP port only and not others. I have updated the PR and docs for it.

**Comment by sid2305** [comment]
Thanks for the review @jtracey93. It totally makes sense to only parameterize the SSH/RDP port only and not others. I have updated the PR and docs for it.

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep [hunk 9]
```diff
@@ -235,6 +235,9 @@ param parTags object = {}
 @sys.description('Set Parameter to true to Opt-out of deployment telemetry.')
 param parTelemetryOptOut bool = false
 
+@sys.description('Define outbound destination ports or ranges for SshRDPOutbound. Default:[22,3389]')
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep:238]:
```suggestion
@sys.description('Define outbound destination ports or ranges for SSH or RDP that you want to access from Azure Bastion.)
```

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep [hunk 10]
```diff
@@ -235,6 +235,9 @@ param parTags object = {}
...
+param parBastionOutboundSshRDPPort array = ['22','3389']
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep:239]:
```suggestion
param parBastionOutboundSshRdpPorts array = ['22','3389']
```
casing and making plural

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep [hunk 11]
```diff
@@ -384,10 +400,7 @@ resource resBastionNsg 'Microsoft.Network/networkSecurityGroups@2021-08-01' = {
           destinationAddressPrefix: 'VirtualNetwork'
           protocol: '*'
           sourcePortRange: '*'
-          destinationPortRanges: [
-            '22'
-            '3389'
-          ]
+          destinationPortRanges: parBastionOutboundSshRDPPort
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep:403]:
```suggestion
          destinationPortRanges: parBastionOutboundSshRdpPorts
```

File: infra-as-code/bicep/modules/hubNetworking/parameters/hubNetworking.parameters.all.json [hunk 12]
```diff
@@ -205,6 +205,9 @@
     },
     "parTelemetryOptOut": {
       "value": false
+    },
+    "parBastionOutboundSshRDPPort": {
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/parameters/hubNetworking.parameters.all.json:209]:
```suggestion
    "parBastionOutboundSshRdpPorts": {
```

File: infra-as-code/bicep/modules/hubNetworking/parameters/mc-hubNetworking.parameters.all.json [hunk 13]
```diff
@@ -167,6 +167,9 @@
     },
     "parTelemetryOptOut": {
       "value": false
+    },
+    "parBastionOutboundSshRDPPort": {
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/parameters/mc-hubNetworking.parameters.all.json:171]:
```suggestion
    "parBastionOutboundSshRdpPorts": {
```

File: infra-as-code/bicep/modules/hubNetworking/generateddocs/hubNetworking.bicep.md [hunk 14]
```diff
@@ -36,6 +36,7 @@ parVpnGatewayConfig | No       | Configuration for VPN virtual network gateway t
 parExpressRouteGatewayConfig | No       | Configuration for ExpressRoute virtual network gateway to be deployed. If a ExpressRoute virtual network gateway is not desired an empty object should be used as the input parameter in the parameter file, i.e. "parExpressRouteGatewayConfig": {   "value": {} }
 parTags        | No       | Tags you would like to be applied to all resources in this module.
 parTelemetryOptOut | No       | Set Parameter to true to Opt-out of deployment telemetry.
+parBastionOutboundSshRDPPort | No       | Define outbound destination ports or ranges for SshRDPOutbound. Default:[22,3389]
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/generateddocs/hubNetworking.bicep.md:39]:
```suggestion
parBastionOutboundSshRdpPorts | No       | Define outbound destination ports or ranges for SSH or RDP that you want to access from Azure Bastion
```

File: infra-as-code/bicep/modules/hubNetworking/generateddocs/hubNetworking.bicep.md [hunk 15]
```diff
@@ -283,6 +284,14 @@ Set Parameter to true to Opt-out of deployment telemetry.
 
 - Default value: `False`
 
+### parBastionOutboundSshRDPPort
+
+![Parameter Setting](https://img.shields.io/badge/parameter-optional-green?style=flat-square)
+
+Define outbound destination ports or ranges for SshRDPOutbound. Default:[22,3389]
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/generateddocs/hubNetworking.bicep.md:291]:
```suggestion
Define outbound destination ports or ranges for SSH or RDP that you want to access from Azure Bastion
```

File: infra-as-code/bicep/modules/hubNetworking/generateddocs/hubNetworking.bicep.md [hunk 16]
```diff
@@ -509,6 +518,12 @@ outHubVirtualNetworkId | string |
         },
         "parTelemetryOptOut": {
             "value": false
+        },
+        "parBastionOutboundSshRDPPort": {
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/generateddocs/hubNetworking.bicep.md:522]:
```suggestion
        "parBastionOutboundSshRdpPorts": {
```

**Review by jtracey93 — CHANGES_REQUESTED** [review]

**Review by jtracey93 — APPROVED** [review]
Looking good, thanks for making the changes.

**Pull request merged**